examples/vmod_event/src/lib.rs

```diff
@@ -1,38 +1,62 @@
-varnish::boilerplate!();
+#[varnish::vmod]
+mod event {
+    use std::sync::atomic::AtomicI64;
+    use std::sync::atomic::Ordering::Relaxed;
 
-use varnish::vcl::ctx::{Ctx, Event, LogTag};
-use varnish::vcl::vpriv::VPriv;
+    use varnish::vcl::ctx::{Ctx, Event, LogTag};
 
-varnish::vtc!(test01);
+    /// number of times on_event() was called with `Event::Load`
+    /// `on_event()` will increment it and store it in the `PRIV_VCL`
+    /// loaded() will access it and return to the VCL
+    static EVENT_LOADED_COUNT: AtomicI64 = AtomicI64::new(0);
 
-// number of times event() was run with Event::Load
-// event will increment it and store a VCL-local copy via VPriv, that loaded() will be able to
-// retrieve
-static mut N: i64 = 0;
+    pub fn loaded(#[shared_per_vcl] shared: Option<&i64>) -> i64 {
+        shared.copied().unwrap_or(0)
+    }
 
-pub fn loaded(_: &Ctx, vp: &VPriv<i64>) -> i64 {
-    // unwrapping is safe here as we know that the event function will have loaded something
-    *vp.as_ref().unwrap()
-}
+    /// FIXME: does Varnish guarantee that `loaded()` will never be called while `on_event()` is running?
+    /// because we are touching a `static mut` variable, this should be highly unsafe, however, Varnish
+    /// guarantees that event functions are called sequentially in the cli thread, so we'll be fine
+    #[event]
+    pub fn on_event(
+        ctx: &mut Ctx,
+        #[shared_per_vcl] shared: &mut Option<Box<i64>>,
+        event: Event,
+    ) -> Result<i64, &'static str> {
+        // log the event, showing that it implements Debug
+        ctx.log(LogTag::Debug, format!("event: {event:?}"));
+
+        // we only care about load events, which is why we don't use `match`
+        if matches!(event, Event::Load) {
+            // increment the count in a thread-safe way
```

Comment:
it doesn't have to be thread-safe, there's only ever one thread reading it. Because the compiler can't prove it doesn't mean it's unsafe. I'm completely fine with escaping with `unsafe` here.

Reply:
First off, most likely this has zero cost on performance - most such constructs are compiled away.

That said, I would recommend against `unsafe` for several reasons:
* it teaches good patterns to possibly-novice users, making them think it is not a big deal to use `unsafe`
* sprinkling `unsafe` effectively makes the entire code unsafe -- I read somewhere a good example of how one unsafe and unsound function would make the rest of the safe rust API be unsound
* user may copy/paste that code elsewhere where they might not understand the implications - resulting in unsound behavior

The whole idea of Rust is that we can code without "thinking" - the compiler does safety checks for us.  Adding unsafe means you have to constantly pay attention to it, which is not common for most Rust devs (that's the who idea of wrapping unsafe API by a small team so that bigger team can treat it as safe).  In short - it would have been safer to write this code in C because at least people are trained to think in terms of safety there.


For now, we still have to use `unsafe` in a few places - lets limit them and try to wrap them cleanly too rather than teaching this.